src/Data/Integer/Divisibility/Properties.agda

```diff
@@ -0,0 +1,197 @@
+
+------------------------------------------------------------------------
+-- _∣_ and _∣′_ are equivalent
+
+∣m⇒∣′m : ∀ {k i} → k ∣ i → k ∣′ i
+∣m⇒∣′m {k} {i} (Ndiv.divides 0 eq) = divides (+ 0) (∣n∣≡0⇒n≡0 eq)
+∣m⇒∣′m {k} {i} (Ndiv.divides q@(ℕ.suc q') eq) with k ≟ + 0
+... | yes refl = divides (+ 0) (∣n∣≡0⇒n≡0 (trans eq (NProp.*-zeroʳ q)))
+... | no ¬k≠0  = divides ((S._*_ on sign) i k ◃ q) (◃-≡ sign-eq abs-eq) where
+
+  k'   = ℕ.suc (ℕ.pred ∣ k ∣)
+  ikq' = sign i S.* sign k ◃ ℕ.suc q'
+
+  sign-eq : sign i ≡ sign (((S._*_ on sign) i k ◃ q) * k)
+  sign-eq = sym $ begin
+    sign (((S._*_ on sign) i k ◃ ℕ.suc q') * k)
+      ≡⟨ cong (λ m → sign (sign ikq' S.* sign k ◃ ∣ ikq' ∣ ℕ.* m))
+              (NProp.m≢0⇒m≡s[pred[m]] (¬k≠0 ∘ ∣n∣≡0⇒n≡0)) ⟩
+    sign (sign ikq' S.* sign k ◃ ∣ ikq' ∣ ℕ.* k')
+      ≡⟨ cong (λ m → sign (sign ikq' S.* sign k ◃ m ℕ.* k'))
+              (abs-◃ (sign i S.* sign k) (ℕ.suc q')) ⟩
+    sign (sign ikq' S.* sign k ◃ _)
+      ≡⟨ sign-◃ (sign ikq' S.* sign k) (ℕ.pred ∣ k ∣ ℕ.+ q' ℕ.* k') ⟩
+    sign ikq' S.* sign k
+      ≡⟨ cong (S._* sign k) (sign-◃ (sign i S.* sign k) q') ⟩
+    sign i S.* sign k S.* sign k
+        ≡⟨ SProp.*-assoc (sign i) (sign k) (sign k) ⟩
+    sign i S.* (sign k S.* sign k)
+      ≡⟨ cong (sign i S.*_) (SProp.s*s≡+ (sign k)) ⟩
+    sign i S.* S.+
+      ≡⟨ SProp.*-identityʳ (sign i) ⟩
+    sign i
+      ∎ where open ≡-Reasoning
+
+  abs-eq : ∣ i ∣ ≡ ∣ ((S._*_ on sign) i k ◃ q) * k ∣
+  abs-eq = sym $ begin
+    ∣ ((S._*_ on sign) i k ◃ ℕ.suc q') * k ∣
+      ≡⟨ abs-◃ (sign ikq' S.* sign k) (∣ ikq' ∣ ℕ.* ∣ k ∣) ⟩
+    ∣ ikq' ∣ ℕ.* ∣ k ∣
+      ≡⟨ cong (ℕ._* ∣ k ∣) (abs-◃ (sign i S.* sign k) (ℕ.suc q')) ⟩
+    ℕ.suc q' ℕ.* ∣ k ∣
+      ≡⟨ sym eq ⟩
+    ∣ i ∣
+      ∎ where open ≡-Reasoning
+
+∣′m⇒∣m : ∀ {k i} → k ∣′ i → k ∣ i
+∣′m⇒∣m {k} {i} (divides q eq) = Ndiv.divides ∣ q ∣ $′ begin
+  ∣ i ∣           ≡⟨ cong ∣_∣ eq ⟩
+  ∣ q * k ∣       ≡⟨ abs-*-commute q k ⟩
+  ∣ q ∣ ℕ.* ∣ k ∣ ∎ where open ≡-Reasoning
+
+------------------------------------------------------------------------
+-- _∣′_ is a preorder
+
+∣′-refl : Reflexive _∣′_
+∣′-refl = ∣m⇒∣′m Ndiv.∣-refl
+
+∣′-reflexive : _≡_ ⇒ _∣′_
+∣′-reflexive refl = ∣′-refl
+
+∣′-trans : Transitive _∣′_
+∣′-trans i∣j j∣k = ∣m⇒∣′m (Ndiv.∣-trans (∣′m⇒∣m i∣j) (∣′m⇒∣m j∣k))
+
+∣′-isPreorder : IsPreorder _≡_ _∣′_
+∣′-isPreorder = record
+  { isEquivalence = isEquivalence
+  ; reflexive     = ∣′-reflexive
+  ; trans         = ∣′-trans
+  }
+
+∣′-preorder : Preorder _ _ _
+∣′-preorder = record { isPreorder = ∣′-isPreorder }
+
+module ∣′-Reasoning = PreorderReasoning ∣′-preorder
+  hiding   (_≈⟨_⟩_)
+  renaming (_∼⟨_⟩_ to _∣′⟨_⟩_)
+
+------------------------------------------------------------------------
+-- Properties of _∣′_
+
+_∣′?_ : Decidable _∣′_
+k ∣′? m = DEC.map′ ∣m⇒∣′m ∣′m⇒∣m (∣ k ∣ Ndiv.∣? ∣ m ∣)
+
+0∣′⇒≡0 : ∀ {m} → + 0 ∣′ m → m ≡ + 0
+0∣′⇒≡0 0|m = ∣n∣≡0⇒n≡0 (Ndiv.0∣⇒≡0 (∣′m⇒∣m 0|m))
+
+m∣′∣m∣ : ∀ {m} → m ∣′ + ∣ m ∣
```

Comment:
The types of these are **really** hard to parse. Might be worth adding a few judicious brackets?